src/uwtools/rocoto.py

```diff
@@ -143,3 +67,276 @@ def validate_rocoto_xml(input_xml: OptionalPath) -> bool:
     for err in list(schema.error_log):
         log.error(err)
     return valid
+
+
+class _RocotoXML:
+    """
+    Generate a Rocoto XML document from a UW YAML config.
+    """
+
+    def __init__(self, config_file: OptionalPath = None) -> None:
+        self._config_validate(config_file)
+        self._config = YAMLConfig(config_file).data
+        self._add_workflow(self._config)
+
+    def dump(self, path: OptionalPath = None) -> None:
+        """
+        Emit Rocoto XML document to file or stdout.
+
+        :param path: Optional path to write XML document to.
+        """
+        # Render the tree to a string, fix mangled entities (e.g. "&amp;FOO;" -> "&FOO;"), insert
+        # !DOCTYPE block, then write final XML.
+        xml = etree.tostring(
+            self._root, pretty_print=True, encoding="utf-8", xml_declaration=True
+        ).decode()
+        xml = re.sub(r"&amp;([^;]+);", r"&\1;", xml)
+        xml = self._insert_doctype(xml)
+        with writable(path) as f:
+            f.write(xml.strip())
+
+    @property
+    def _doctype(self) -> Optional[str]:
+        """
+        Generate the <!DOCTYPE> block with <!ENTITY> definitions.
+
+        :return: The <!DOCTYPE> block if entities are defined, otherwise None.
+        """
+        if entities := self._config[STR.workflow].get(STR.entities):
+            tags = (f'  <!ENTITY {k} "{v}">' for k, v in entities.items())
```

Comment:
I agree it would be better to be consistent -- especially since it was not immediately clear to you what these variables meant. Often, in these kinds of concise expressions, one-letter iterator variables are helpful, but I think consistency trumps conciseness here. Will fix.